src/canbus/elm327.cpp

```diff
@@ -0,0 +1,187 @@
+#include "canbus/elm327.hpp"
+
+elm327::elm327(QString canInterface)
+{
+    DASH_LOG(info)<<"[ELM327] Connecting elm";
+    this->connect(canInterface, B115200);
+    if (this->connected) this->initialize();
+
+}
+elm327::~elm327()
+{
+    this->raw_query("ATZ");
+    sleep(1);
+    close(this->fd);
+    this->connected = false;
+}
+
+elm327 *elm327::get_instance()
+{
+    static elm327 elm;
+    return &elm;
+}
+
+int elm327::_write(std::string str)
+{
+    str += '\r';
+    int size;
+    if ((size = write(this->fd, str.c_str(), str.length())) < 0) {
+        DASH_LOG(error) << "[ELM327] failed write" << std::endl;
+        this->connected = false;
+        return 0;
+    }
+
+    return size;
+}
+
+
+void elm327::connect(QString dev_path, speed_t baudrate)
+{
+    if (this->connected) return;
+
+    this->fd = open(dev_path.toStdString().c_str(), O_RDWR | O_NOCTTY | O_NDELAY);
+    if (this->fd == -1) {
+        this->connected = false;
+        return;
+    }
+
+    fcntl(this->fd, F_SETFL, 0);
+
+    struct termios options;
+    tcgetattr(this->fd, &options);
+
+    options.c_cflag |= (CLOCAL | CREAD);
+    options.c_lflag &= !(ICANON | ECHO | ECHOE | ISIG);
+    options.c_oflag &= !(OPOST);
+    options.c_cc[VMIN] = 0;
+    options.c_cc[VTIME] = 100;
+
+    tcflush(this->fd, TCIOFLUSH);
+    tcsetattr(this->fd, TCSANOW, &options);
+
+    cfsetispeed(&options, baudrate);
+    cfsetospeed(&options, baudrate);
+    tcsetattr(this->fd, TCSANOW, &options);
+    tcflush(this->fd, TCIOFLUSH);
+
+    this->connected = true;
+}
+void elm327::initialize()
+{
+    DASH_LOG(info)<<"[ELM327] Initializing elm";
+
+    std::vector<std::string> cmds = {"ATD", "ATZ", "ATE0", "ATL0", "ATS0", "AT H0", "ATSP0", "0100"};
+    for (auto const cmd : cmds) {
+        this->raw_query(cmd);
+        usleep(500000);
+    }
+}
+
+
+bool elm327::writeFrame(QCanBusFrame frame)
+{
+    if(!this->connected) return false;
+    std::lock_guard<std::mutex> guard(this->elm_mutex);
+    std::stringstream ss;
+        
+    //this is an obd message, so we can send it with the elm327
+    if(frame.frameId() == 0x7df){
```

Comment:
`0x7df` represents an OBD-II request message. That is not obvious from the code/comments. As a matter of style and readability, consider naming your magic numbers with `#define` or `enum`